homeassistant/components/heos/media_player.py

```diff
@@ -0,0 +1,175 @@
+"""Denon HEOS Media Player."""
+
+from homeassistant.components.media_player import MediaPlayerDevice
+from homeassistant.components.media_player.const import (
+    DOMAIN, MEDIA_TYPE_MUSIC, SUPPORT_NEXT_TRACK, SUPPORT_PAUSE, SUPPORT_PLAY,
+    SUPPORT_PLAY_MEDIA, SUPPORT_PREVIOUS_TRACK, SUPPORT_STOP,
+    SUPPORT_VOLUME_MUTE, SUPPORT_VOLUME_SET, SUPPORT_VOLUME_STEP)
+from homeassistant.const import STATE_OFF, STATE_PAUSED, STATE_PLAYING
+from homeassistant.helpers.dispatcher import (
+    async_dispatcher_connect, async_dispatcher_send)
+
+from . import DOMAIN as HEOS_DOMAIN
+
+DEPENDENCIES = ['heos']
+
+SUPPORT_HEOS = SUPPORT_PLAY | SUPPORT_STOP | SUPPORT_PAUSE | \
+        SUPPORT_PLAY_MEDIA | SUPPORT_PREVIOUS_TRACK | SUPPORT_NEXT_TRACK | \
+        SUPPORT_VOLUME_MUTE | SUPPORT_VOLUME_SET | SUPPORT_VOLUME_STEP
+
+SIGNAL_HEOS_UPDATE = 'heos_update'
+
+PLAY_STATE_TO_STATE = {
+    'play': STATE_PLAYING,
+    'pause': STATE_PAUSED,
+    'stop': STATE_OFF
+}
+
+
+async def async_setup_platform(hass, config, async_add_devices,
+                               discover_info=None):
+    """Set up the HEOS platform."""
+    players = hass.data[HEOS_DOMAIN][DOMAIN]
+    devices = [HeosMediaPlayer(p) for p in players]
+    async_add_devices(devices, True)
+
+
+class HeosMediaPlayer(MediaPlayerDevice):
+    """The HEOS player."""
+
+    def __init__(self, player):
+        """Initialize."""
+        self._player = player
+        self._dispatcher_remove = None
+        self._player.state_change_callback = self._update_state
+
+    def _update_state(self):
+        async_dispatcher_send(self.hass, SIGNAL_HEOS_UPDATE, self.unique_id)
```

Comment:
If we're only signaling within this entity we don't need the dispatch helper at all.

Just `await self.async_update_ha_state()` here.

Reply:
Or rather call `self.async_schedule_update_ha_state` since we're not in a coroutine.

Reply:
Yeah I suggested dispatcher to solve the sequencing of the entity being added to HASS vs potential for callbacks.  Moving the registration as per your other comment solves that much better.  :man_facepalming: